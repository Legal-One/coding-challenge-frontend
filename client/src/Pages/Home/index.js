import React, { useState, useEffect } from "react";
import axios from "axios";
import CallLogs from "../../components/CallLogs";

const apiUrl = "http://localhost:5000/api/call-logs";
function Home() {
    const [callLogs, setCallLogs] = useState([]);

    useEffect(() => {
        const fetchCallLogs = async () => {
            const response = await axios.get(apiUrl);
            setCallLogs(response.data);
        };
        fetchCallLogs();
    }, []);

    return (
        <div>
            <CallLogs callLogs={callLogs} />
        </div>
    );
}
export default Home;
