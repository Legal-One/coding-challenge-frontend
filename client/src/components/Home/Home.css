table {
  font-family: "Poppins";
  color: #454545;
  border-collapse: collapse;
  width: 100%;
  margin-bottom: 5em;
}

.table-holder {
  overflow-x: auto;
}

td,
th {
  border: 1px solid #dddddd;
  text-align: left;
  padding: 8px 14px;
}

tr {
  height: 60px;
}

tr:nth-child(even) {
  background-color: #dddddd;
}

.agent-name {
  cursor: pointer;
}

.home .icons {
  display: flex;
  margin-bottom: 1em;
  justify-content: flex-end;
}

.icons img {
  height: 20px;
  margin-left: 2em;
  cursor: pointer;
}

@media (max-width: 655px) {
  table {
    width: 45em;
  }
}
