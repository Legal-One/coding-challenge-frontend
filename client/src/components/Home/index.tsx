import React, { useEffect, useCallback } from 'react'
import Table from '../Table'
import {
  getCalls,
  selectLoadStatus,
  selectCallDetails
} from '../../features/callcenter/callcenter-slice'
import { useAppDispatch } from '../../store'
import { useSelector } from 'react-redux'
import { LoadStatus } from '../../types'
import ResolutionGraph from '../ResolutionGraph'
import LineGraph from '../LineGraph'
import TableHeader from '../Table/table-header'
import TableRow from '../Table/call-statistics-table-row'

function Home() {
  const tableHeaders = ['Phone number', 'Number of calls', 'Last Spoken Agent', 'Last call time']
  const dispatch = useAppDispatch()

  const getCallData = useCallback(async () => {
    console.log('It is called')
    await dispatch(getCalls())
  }, [dispatch])

  const loadStatus: LoadStatus = useSelector(selectLoadStatus)

  useEffect(() => {
    if (loadStatus === LoadStatus.NOTLOADED) getCallData()
  }, [getCallData, loadStatus])

  const callStats = useSelector(selectCallDetails)

  return (
    <div>
      <h1 className="is-size-2 has-text-weight-bold pt-6 pb-6 has-text-centered">
        Call Details at a Glance
      </h1>
      <div className="container is-flex is-justify-content-space-between">
        <div className="card">
          <span>Total Number of Calls</span>
          <div className="is-size-2 has-text-weight-bold">{callStats.total.calls}</div>
        </div>

        <div className="card">
          <span>Average Call Time</span>
          <div className="is-size-2 has-text-weight-bold">
            {callStats.total.calls > 0 &&
              (callStats.total.duration / callStats.total.calls).toFixed(2)}
            secs
          </div>
        </div>
        <div className="card">
          <span>Unique Numbers Called</span>
          <div className="is-size-2 has-text-weight-bold">{Object.keys(callStats).length - 1}</div>
        </div>
      </div>
      <div className="is-flex is-justify-content-center">
        <ResolutionGraph />
        <LineGraph />
      </div>
      <Table>
        <TableHeader headers={tableHeaders} />
        {Object.keys(callStats).map((key, index) => {
          return (
            key !== 'total' && (
              <TableRow
                key={key}
                {...callStats[key]}
                isLastRow={index === Object.keys(callStats).length - 1}
              />
            )
          )
        })}
      </Table>
    </div>
  )
}

export default Home
