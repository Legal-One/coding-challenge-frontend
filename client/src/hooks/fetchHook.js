import { useState, useEffect } from "react";
import axios from "axios";

export const useFetchData = (apiUrl) => {
    const [data, setData] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);

    useEffect(() => {
        const fetchData = async () => {
            setIsError(false);
            setIsLoading(true);

            try {
                const result = await axios(apiUrl);
                setData(result.data);
            } catch (error) {
                setIsError(true);
            }

            setIsLoading(false);
        };

        fetchData();
    }, [apiUrl]);

    return [data, isLoading, isError];
};
