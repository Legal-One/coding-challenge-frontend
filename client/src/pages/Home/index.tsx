import React from 'react'
import { useSelector } from 'react-redux'

import MainTable from '../../components/MainTable'
import { AppState } from '../../Types'

const loadStyle = {
  margin: 'auto',
  width: '50%',
}

const Home = () => {
  const loading = useSelector((state: AppState) => state.logs.loading)

  return (
    <>
      {loading ? (
        <h1 style={loadStyle}>Loading...</h1>
      ) : (
        <table>
          <tbody>
            <tr>
              <th>Phone number</th>
              <th>Number of calls</th>
              <th>Last call details </th>
            </tr>
          </tbody>
          <tbody>
            <MainTable />
          </tbody>
        </table>
      )}
    </>
  )
}

export default Home
