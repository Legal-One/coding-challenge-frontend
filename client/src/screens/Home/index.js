import React from 'react'
import MapComponent from '../../components/MapComponent';
import Table from '../../components/Table';
import { convetSecToTime, montlyCallLogs } from '../../utils';
import { useHistory } from 'react-router-dom';
import { MAP_AXIS, MAP_TITLE, TABLE_HEADINGS } from '../../utils/constants';

function Home({ logs, agents }) {

    const componentData = montlyCallLogs({ logs, agents })
    const history = useHistory();

    return (
        <div>
            <MapComponent data={componentData.mapData} axis={MAP_AXIS.HOME} title={MAP_TITLE.HOME} />
            <Table headings={TABLE_HEADINGS.HOME} >
                {
                    componentData.tableData &&
                    Object.values(componentData.tableData).map((val) => (
                        <tr key={val[0].number}>
                            <td><span onClick={() => history.push(`/call/${val[0].number}`)}>
                                {val[0].number}
                            </span></td>
                            <td>{val.length}</td>
                            <td>
                                <span onClick={() => history.push(`/agent/${val[0].agentIdentifier}`)}> {val[0].agentName} </span> / {`${convetSecToTime(val[0].duration)}`}
                            </td>
                        </tr>
                    ))
                }
            </Table>
        </div>
    )
}

export default Home;
