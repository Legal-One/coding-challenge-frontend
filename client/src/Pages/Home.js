import { Link } from "react-router-dom";
import { useFetchData } from "../hooks/fetchHook";
import Table from "../components/Table";
import Chart from "../components/Chart";
import { getFormattedTime } from "../utils/timeFormat";

import "./index.scss";

const apiUrl = "http://localhost:5000/api/call-logs";
const headers = ["Phone number", "Number of calls", "Last call details"];

const dataMapper = (cl) => {
    return cl.map((callLog) => {
        return [
            <Link to={`/call/${callLog.phoneNumber}`}>
                {callLog.phoneNumber}
            </Link>,
            callLog.callCount + (callLog.callCount === 1 ? " call" : " calls"),
            <>
                <Link to={`/agent/${callLog.agentIdentifier}`}>
                    {callLog.agentName}
                </Link>
                &nbsp;/&nbsp;{getFormattedTime(callLog.lastCallTime)}
            </>,
        ];
    });
};

function Home() {
    const [data, isLoading, isError] = useFetchData(apiUrl);

    const chartData = [
        ["Phone number", "Call count"],
        ...data.map((cl) => [cl.phoneNumber, parseInt(cl.callCount)]),
    ];

    return (
        <main>
            {!isLoading && (
                <>
                    <Table headers={headers} rows={dataMapper(data)} />
                    <Chart chartType="BarChart" chartData={chartData} />
                </>
            )}
            {isLoading && <h3>Fetching data...</h3>}
            {isError && <h3>Error fetching data. Please refresh your page</h3>}
        </main>
    );
}
export default Home;
