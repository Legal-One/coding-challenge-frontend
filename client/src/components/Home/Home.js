import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import { getAccumulatedLogs } from "../../api";
import "./Home.css";

const Home = () => {
  const [logs, setLogs] = useState([]);
  const fetchLogs = async () => {
    const logs = await getAccumulatedLogs();
    if (logs.length) {
      return setLogs(logs);
    }
  };

  useEffect(() => fetchLogs(), []);
  return (
    <div className="home">
      <table>
        <thead>
          <tr>
            <th>Phone number</th>
            <th>Number of calls</th>
            <th>Agent name</th>
          </tr>
        </thead>
        <tbody>
          {logs.map((log) => (
            <tr key={log.identifier}>
              <th>
                <Link to={`/call/${log.number}`}>{log.number}</Link>
              </th>
              <th>{log.count}</th>
              <th className="agent-name">
                <Link to={`/agent/${log.agent.identifier}`}>
                  {" "}
                  {log.agent.firstName} {log.agent.lastName}
                </Link>
              </th>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Home;
