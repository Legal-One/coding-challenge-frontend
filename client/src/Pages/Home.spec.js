import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import * as mockHook from "../hooks/fetchHook";
import Home from "./Home";

describe("HomePage", () => {
    test("render with all props passed", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [
                {
                    agentName: "John Doe",
                    agentIdentifier: "beb04834-9f8b-11e7-9469-8a39b455b609",
                    lastCallTime: "2020-10-06T13:50:00.000Z",
                    phoneNumber: "+4912345678",
                    callCount: 13,
                },
                {
                    agentName: "Spiderman",
                    agentIdentifier: "aeb04834-9f8b-11e7-9469-8a39b455b709",
                    lastCallTime: "2020-11-06T13:50:00.000Z",
                    phoneNumber: "+4912345678",
                    callCount: 1,
                },
            ],
            false,
            false,
        ]);
        const { asFragment } = render(
            <Router>
                <Home />
            </Router>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test("render with error from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            false,
            true,
        ]);
        const { asFragment } = render(
            <Router>
                <Home />
            </Router>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    test("render when waiting for response from api service", () => {
        jest.spyOn(mockHook, "useFetchData").mockImplementation(() => [
            [],
            true,
            false,
        ]);
        const { asFragment } = render(
            <Router>
                <Home />
            </Router>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
