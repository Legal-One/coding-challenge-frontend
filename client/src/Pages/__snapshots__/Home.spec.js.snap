// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage render when waiting for response from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Fetching data...
    </h3>
  </main>
</DocumentFragment>
`;

exports[`HomePage render with all props passed 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Call summary
    </h3>
    <table
      class="Table"
    >
      <thead>
        <tr>
          <th>
            Phone number
          </th>
          <th>
            Number of calls
          </th>
          <th>
            Last call details
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/call/+4912345678"
            >
              +4912345678
            </a>
          </td>
          <td>
            13 calls
          </td>
          <td>
            <a
              href="/agent/beb04834-9f8b-11e7-9469-8a39b455b609/John Doe"
            >
              John Doe
            </a>
             / 15:50
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/call/+4912345678"
            >
              +4912345678
            </a>
          </td>
          <td>
            1 call
          </td>
          <td>
            <a
              href="/agent/aeb04834-9f8b-11e7-9469-8a39b455b709/Spiderman"
            >
              Spiderman
            </a>
             / 14:50
          </td>
        </tr>
      </tbody>
    </table>
  </main>
</DocumentFragment>
`;

exports[`HomePage render with error from api service 1`] = `
<DocumentFragment>
  <main>
    <h3>
      Error fetching data. Please refresh your page
    </h3>
  </main>
</DocumentFragment>
`;
